### PR TITLE
Refactor `lib/files_list.js`

### PR DESCRIPTION
**Pull request by Dignifiedquire**
As noted in multiple places `lib/file_list.js` is in need for a refactor and cleanup. This is that.
At the moment very much a work in progress

- [x] Ensure `refresh` is not triggered while refreshing
- [x] `reload`
- [x] `addFile`
- [x] `changeFile`
- [x] `removeFile`
- [x] `_refresh`
- [x] finish porting all old tests
- [x] fix tests
- [x] cleanup

Ref #1447 #1436

### CODE REVIEW

**Comment by Dignifiedquire** [comment]
I signed it!

File: lib/file-list.js [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+          log.debug('Excluded file "%s"', path)
+          return Promise.resolve()
+        }
+
+        var mtime = mg.statCache[path].mtime
+        var doNotCache = patternObject.nocache
+        var file = new File(path, mtime, doNotCache)
+
+        if (file.doNotCache) {
+          log.debug('Not preprocessing "s" due to nocache')
```
Comment (nmalaguti) [inline lib/file-list.js]: Missing %

**Comment by zzo** [comment]
how's this coming along?  once you've got it in decent shape I'd like to start testing it with angular thanks!

**Comment by Dignifiedquire** [comment]
I made some progress today. Hopefully I'll be able to finish it some time this week

**Comment by Dignifiedquire** [comment]
@zzo @maksimr I think this is ready, all tests pass, and the bug in #1436 is fully fixed for me. Can you guys give this a sanity check please before I merge

File: lib/file-list.js [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+var Url = require('./url')
+var helper = require('./helper')
+var _ = helper._
+var log = require('./logger').create('watcher')
+
+// Constants
+// ---------
+
+var GLOB_OPTS = {
+  // globDebug: true,
```
Comment (nmalaguti) [inline lib/file-list.js]: Should this commented line be deleted?

**Comment by nmalaguti** [comment]
Great work, the refactor looks very clean and much easier to follow.

**Comment by Dignifiedquire** [comment]
Thanks @nmalaguti

**Pull request merged**